### ISSUE TITLE
Semicolon rule should flag unnecessary semicolons on class member functions

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#1476]
We changed the behavior of the semicolon rule to allow omission of the trailing semicolon after bound class member functions:

``` ts
class Foo {
  public renderFoo = () => {
    return null;
  }
   ^ right here
}
```

I propose that we actually _ban_ semicolons at that location in TSLint 4.0. We can support the old functionality with a new option `allow-bound-class-methods`.

**Comment by JKillian** [comment]
I'm not convinced this is need for 4.0, but wouldn't hurt to get it in either. It does seem like a little bit of a minute option though, I'm not entirely convinced it's worth the added complexity

**Comment by mprobst** [comment]
Just as a feedback item (not sure if this is the right venue?), we just stumbled across this, and were very surprised this.

I probably don't quite understand the rationale behind this, but in particular that tslint flags semicolons as *unnecessary* in this position (i.e. doesn't just accept leaving them out) seems confusing.

**Comment by adidahiya** [comment]
You can get the old behavior back with the `ignore-bound-class-methods` option. We wanted to be a little more opinionated with the default configuration and make bound class methods look like regular class methods in most TS code.

**Comment by mprobst** [comment]
Yeah, I figured so much. I'd have expected `tslint` to out-of-the-box require that semicolon, erring more on the side of consistency, or ignoring the issue, but adding that option back in works for us.

**Comment by bertyhell** [comment]
I tried to get this to stop throwing errors in intellij

I tried removing the semicolon behind class arrow functions, then i get this warning:
![image](https://cloud.githubusercontent.com/assets/1710840/22502847/1a2d3e44-e86f-11e6-80cb-e1c7d6808794.png)

I tried enabling the option like this:
![image](https://cloud.githubusercontent.com/assets/1710840/22502855/24483406-e86f-11e6-8490-04d96c5d022e.png)

But i get this result:
![image](https://cloud.githubusercontent.com/assets/1710840/22502859/2c69aca0-e86f-11e6-904a-99055dba5ae2.png)

I tried enabling the option like this:
![image](https://cloud.githubusercontent.com/assets/1710840/22502868/352c6d1e-e86f-11e6-9302-c82d839d6648.png)

But then it stops checking for semicolons all together.

tslint version: 4.4.2
ts version: 2.1.5
intellij version: 2016.3
OS: windows 10 x64

**Comment by popstr** [comment]
For people coming to this thread looking for the correct tslint config (as I did) - the correct syntax for this would be:
"semicolon": [true, "always", "ignore-bound-class-methods"]

See: https://palantir.github.io/tslint/rules/semicolon/

**Comment by robertpenner** [comment]
@popstr Yes, that's what I'm using as well.
```
"semicolon": [true, "always", "ignore-bound-class-methods"]
```

**Comment by aaronbeall** [comment]
Using version `5.2.0` when I try `"semicolon": [true, "always", "ignore-bound-class-methods"]` I still get errors on:

```
class Foo {
  public bar = () => {
  };
}
```

> ERROR: src/Foo.ts[22, 4]: Unnecessary semicolon

**Comment by rkrisztian** [comment]
I've created #3216 because ignoring is a bad idea in terms of consistency, and the rule without the `ignore-bound-class-methods` option is **still buggy** as I found out.
